designs/cloudstack-multiple-endpoints.md

```diff
@@ -49,50 +49,50 @@ will need to be available on *all* the Cloudstack API endpoints. We will validat
-	Zones []CloudStackZone `json:"zones"`
-	// Account typically represents a customer of the service provider or a department in a large organization. Multiple users can exist in an account, and all CloudStack resources belong to an account. Accounts have users and users have credentials to operate on resources within that account. If an account name is provided, a domain must also be provided.
-	Account string `json:"account,omitempty"`
-	// CloudStack Management API endpoint's IP. It is added to VM's noproxy list
-        ManagementApiEndpoint string `json:"managementApiEndpoint"`
+    // Domain contains a grouping of accounts. Domains usually contain multiple accounts that have some logical relationship to each other and a set of delegated administrators with some authority over the domain and its subdomains
+    Domain string `json:"domain"`
+    // Zones is a list of one or more zones that are managed by a single CloudStack management endpoint.
+    Zones []CloudStackZone `json:"zones"`
+    // Account typically represents a customer of the service provider or a department in a large organization. Multiple users can exist in an account, and all CloudStack resources belong to an account. Accounts have users and users have credentials to operate on resources within that account. If an account name is provided, a domain must also be provided.
+    Account string `json:"account,omitempty"`
+    // CloudStack Management API endpoint's IP. It is added to VM's noproxy list
+    ManagementApiEndpoint string `json:"managementApiEndpoint"`
 }
 ```
 
 We would instead propose to gradually deprecate all the existing attributes and instead, simply include a list of AvailabilityZone objects like so
 
 ```
 type CloudStackDatacenterConfigSpec struct {
-	// Deprecated
-	Domain string `json:"domain,omitempty"`
-	// Deprecated
-	Zones []CloudStackZone `json:"zones,omitempty"`
-	// Deprecated
-	Account string `json:"account,omitempty"`
-	// Deprecated
-	ManagementApiEndpoint string `json:"managementApiEndpoint,omitempty"`
-	// List of AvailabilityZones to distribute VMs across - corresponds to a list of CAPI failure domains
-	AvailabilityZones []CloudStackAvailabilityZone `json:"availabilityZones,omitempty"`
+    // Deprecated
+    Domain string `json:"domain,omitempty"`
+    // Deprecated
+    Zones []CloudStackZone `json:"zones,omitempty"`
+    // Deprecated
+    Account string `json:"account,omitempty"`
+    // Deprecated
+    ManagementApiEndpoint string `json:"managementApiEndpoint,omitempty"`
+    // List of AvailabilityZones to distribute VMs across - corresponds to a list of CAPI failure domains
+    AvailabilityZones []CloudStackAvailabilityZone `json:"availabilityZones,omitempty"`
 }
 ```
 
 where each AvailabilityZone object looks like
 
 ```
 type CloudStackAvailabilityZone struct {
-	// Name would be used to match the availability zone defined in the datacenter config to the credentials passed in from the cloud-config ini file
```

Comment:
I would suggest to keep this unless there is a mayor concern
We learnt with the worker node groups that it makes comparing arrays easier (which is helpful to detect updates, removal, etc.)

Reply:
We can maybe default the names? I would argue that if users name they would be able to properly associate which one is related to the appropriate credentialref and all other fields, especially because we are not going with the CRD approach, which would anyways require a name. So I would also vote for a name unless we have a strong reason not to.

Reply:
Noted - will re-add and autogenerate. We can also add a validation step to ensure the names are unique